group.c

```diff
@@ -77,26 +97,30 @@ static int empty_group(struct Group *g)
   return !g->as && !g->rs;
 }
 
-void mutt_group_context_add(struct GroupContext **ctx, struct Group *group)
+void mutt_group_context_add(struct GroupContextHead *head, struct Group *group)
 {
-  for (; *ctx; ctx = &((*ctx)->next))
+  struct GroupContext *np = NULL;
+  STAILQ_FOREACH(np, head, entries)
   {
-    if ((*ctx)->g == group)
+    if (np->g == group)
       return;
   }
-
-  *ctx = mutt_mem_calloc(1, sizeof(struct GroupContext));
-  (*ctx)->g = group;
+  np = mutt_mem_calloc(1, sizeof(struct GroupContext));
+  np->g = group;
+  STAILQ_INSERT_TAIL(head, np, entries);
 }
 
-void mutt_group_context_destroy(struct GroupContext **ctx)
+/* Double check */
+void mutt_group_context_destroy(struct GroupContextHead *head)
 {
-  struct GroupContext *p = NULL;
-  for (; *ctx; *ctx = p)
+
+  struct GroupContext *np = NULL, *tmp = NULL;
+  STAILQ_FOREACH_SAFE(np, head, entries, tmp)
   {
-    p = (*ctx)->next;
-    FREE(ctx);
+    STAILQ_REMOVE(head, np, GroupContext, entries);
+    FREE(&np);
```

Comment:
Please use the pattern used in mutt_list_free in mutt/list.c for a quicker removal of all elements in the list.